packages/@react-aria/autocomplete/src/useAutocomplete.ts

```diff
@@ -0,0 +1,278 @@
+  let delayNextActiveDescendant = useRef(false);
+  let lastCollectionNode = useRef<HTMLElement>(null);
+
+  let updateActiveDescendant = useEffectEvent((e) => {
+    let {target} = e;
+    clearTimeout(timeout.current);
+    e.stopPropagation();
+
+    if (target !== collectionRef.current) {
+      if (delayNextActiveDescendant.current) {
+        timeout.current = setTimeout(() => {
+          state.setFocusedNodeId(target.id);
+        }, 500);
+      } else {
+        state.setFocusedNodeId(target.id);
+      }
+    } else {
+      state.setFocusedNodeId(null);
+    }
+
+    delayNextActiveDescendant.current = false;
+  });
+
+  let callbackRef = useCallback((collectionNode) => {
+    if (collectionNode != null) {
+      // When typing forward, we want to delay the setting of active descendant to not interrupt the native screen reader announcement
+      // of the letter you just typed. If we recieve another UPDATE_ACTIVEDESCENDANT call then we clear the queued update
+      // We track lastCollectionNode to do proper cleanup since callbackRefs just pass null when unmounting. This also handles
+      // React 19's extra call of the callback ref in strict mode
+      lastCollectionNode.current?.removeEventListener(UPDATE_ACTIVEDESCENDANT, updateActiveDescendant);
+      lastCollectionNode.current = collectionNode;
+      collectionNode.addEventListener(UPDATE_ACTIVEDESCENDANT, updateActiveDescendant);
+    } else {
+      lastCollectionNode.current?.removeEventListener(UPDATE_ACTIVEDESCENDANT, updateActiveDescendant);
+    }
+  }, [updateActiveDescendant]);
+
+  // Make sure to memo so that React doesn't keep registering a new event listeners on every rerender of the wrapped collection
+  let mergedCollectionRef = useObjectRef(useMemo(() => mergeRefs(collectionRef, callbackRef), [collectionRef, callbackRef]));
+
+  let focusFirstItem = useEffectEvent(() => {
+    delayNextActiveDescendant.current = true;
+    collectionRef.current?.dispatchEvent(
+      new CustomEvent(FOCUS_EVENT, {
+        cancelable: true,
+        bubbles: true,
+        detail: {
+          focusStrategy: 'first'
+        }
+      })
+    );
+  });
+
+  let clearVirtualFocus = useEffectEvent(() => {
+    state.setFocusedNodeId(null);
+    let clearFocusEvent = new CustomEvent(CLEAR_FOCUS_EVENT, {
+      cancelable: true,
+      bubbles: true
+    });
+    clearTimeout(timeout.current);
+    delayNextActiveDescendant.current = false;
+    collectionRef.current?.dispatchEvent(clearFocusEvent);
+  });
+
+  // Tell wrapped collection to focus the first element in the list when typing forward and to clear focused key when deleting text
+  // for screen reader announcements
+  let lastInputValue = useRef<string | null>(null);
+  useEffect(() => {
+    if (state.inputValue != null) {
+      if (lastInputValue.current != null && lastInputValue.current !== state.inputValue && lastInputValue.current?.length <= state.inputValue.length) {
```

Comment:
what if you copy and pasted? the value might be completely different but also shorter which isn't the same as backspacing I guess. what do you think should happen in that case?

Reply:
I think focus should be cleared in that case since the announcement would be a longer one (announcing the new word). TBH that should also apply when pasting a string that is longer that your current too but currently will autofocus the first option... In a ideal world screen readers would be better about not interrupting these announcements so we could just always autofocus haha. I guess I could try and add some logic to differentiate typing forward vs copy pasting by comparing just how much the input value has changed from the last value

Reply:
feels pretty gross that we'd be diffing the strings to see if it is a different word + checking the lengths just to make it not autofocus when pasting a longer/different string in the field. Perhaps we could make an assumption that a screenreader user would already know what string they are copy and pasting and make the case that interrupted announcement isn't super important?

Reply:
yeah, the native `input` event provides an `inputType` property that has all sorts of "reasons" for the change, which might be useful. We use the related `beforeinput` event in number field. I know you mentioned you tried onChange and it was happening before filtering, but maybe we could use that event to track the change type and apply it after the next render. Can be a future enhancement.

Reply:
sounds good, I'll make a follow up ticket to track and do some initial tinkering to see if this is feasible (thanks for the tip with `inputType`, forgot that was a thing)